### ISSUE TITLE
Reset/Reinitialize model weights/parameters

### ISSUE THREAD

**Issue by NEGU93** [tensorflow/tensorflow#48230]
Motivated by [this](https://github.com/keras-team/keras/issues/341) post, I would like to ask for a method like `model.reset_weights()` to get the reset and init the weights again.

**Comment by fchollet** [comment]
Hi there,

You can straightforwardly roll out your own function. Like this:

```python
def reinitialize_model(model):
  weights = []
  initializers = []
  for layer in model.layers:
    if isinstance(layer, (keras.layers.Dense, keras.layers.Conv2D)):
      weights += [layer.kernel, layer.bias]
      initializers += [layer.kernel_initializer, layer.bias_initializer]
    elif isinstance(layer, keras.layers.BatchNormalization):
      weights += [layer.gamma, layer.beta, layer.moving_mean, layer.moving_variance]
      initializers += [layer.gamma_initializer,
                       layer.beta_initializer,
                       layer.moving_mean_initializer,
                       layer.moving_variance_initializer]
  for w, init in zip(weights, initializers):
    w.assign(init(w.shape, dtype=w.dtype))

reinitialize_model(model)
```

You will need to add coverage for all the layers (with weights) you use in your model.

**Comment by edubergeek** [comment]
Merci

**Comment by charleswilmot** [comment]
Not so straightforward

**Comment by unagimaster** [comment]
What do you mean by "add coverage for all layers"? @fchollet

**Comment by OhnkytaBlabdey** [comment]
Does this solved by anyone?
I have Bi-LSTM in my model, so every time I construct a new instance(but the old one's gpu memory can not be released)

**Comment by phrasenmaeher** [comment]
I've worked with the suggested code a bit; here's my improved version. It filters `None` values and allows to use any passed initializer, overriding the layer's default one.
```python
weights = []
initializers = []


def reinitialize_model(model, new_initializer: tf.keras.initializers.Initializer = None):
    def remove_nones(weights, initializers):
        w = []
        i = []
        for element, initializer in zip(weights, initializers):
            if element is not None:
                w.append(element)
                i.append(initializer)
        return w, i

    global weights, initializers

    for layer in model.layers:
        if isinstance(layer, (keras.layers.Dense, keras.layers.Conv2D)):
            weights += [layer.kernel, layer.bias]
            initializers += [layer.kernel_initializer, layer.bias_initializer]
        elif isinstance(layer, keras.layers.BatchNormalization):
            weights += [layer.gamma, layer.beta, layer.moving_mean, layer.moving_variance]
            initializers += [layer.gamma_initializer,
                             layer.beta_initializer,
                             layer.moving_mean_initializer,
                             layer.moving_variance_initializer]
        elif isinstance(layer, (keras.layers.InputLayer,
                                keras.layers.Reshape,
                                keras.layers.Concatenate,
                                keras.layers.ZeroPadding2D,
                                keras.layers.MaxPool2D,
                                keras.layers.Activation,
                                keras.layers.Add,
                                keras.layers.GlobalAveragePooling2D,
                                keras.layers.Flatten

                                )):
            # These layers don't need initialization
            continue
        else:
            raise ValueError('Unhandled layer type: %s' % (type(layer)))

    if new_initializer:
        initializers = [new_initializer for _ in initializers]
    weights, initializers = remove_nones(weights, initializers)
    for w, init in zip(weights, initializers):
        w.assign(init(w.shape, dtype=w.dtype))
```

**Comment by dubovikmaster** [comment]
> Я немного поработал с предложенным кодом; вот моя улучшенная версия. Он фильтрует `None`значения и позволяет использовать любой переданный инициализатор, переопределяя инициализатор слоя по умолчанию.
> 
> ```python
> weights = []
> initializers = []
> 
> 
> def reinitialize_model(model, new_initializer: tf.keras.initializers.Initializer = None):
>     def remove_nones(weights, initializers):
>         w = []
>         i = []
>         for element, initializer in zip(weights, initializers):
>             if element is not None:
>                 w.append(element)
>                 i.append(initializer)
>         return w, i
> 
>     global weights, initializers
> 
>     for layer in model.layers:
>         if isinstance(layer, (keras.layers.Dense, keras.layers.Conv2D)):
>             weights += [layer.kernel, layer.bias]
>             initializers += [layer.kernel_initializer, layer.bias_initializer]
>         elif isinstance(layer, keras.layers.BatchNormalization):
>             weights += [layer.gamma, layer.beta, layer.moving_mean, layer.moving_variance]
>             initializers += [layer.gamma_initializer,
>                              layer.beta_initializer,
>                              layer.moving_mean_initializer,
>                              layer.moving_variance_initializer]
>         elif isinstance(layer, (keras.layers.InputLayer,
>                                 keras.layers.Reshape,
>                                 keras.layers.Concatenate,
>                                 keras.layers.ZeroPadding2D,
>                                 keras.layers.MaxPool2D,
>                                 keras.layers.Activation,
>                                 keras.layers.Add,
>                                 keras.layers.GlobalAveragePooling2D,
>                                 keras.layers.Flatten
> 
>                                 )):
>             # These layers don't need initialization
>             continue
>         else:
>             raise ValueError('Unhandled layer type: %s' % (type(layer)))
> 
>     if new_initializer:
>         initializers = [new_initializer for _ in initializers]
>     weights, initializers = remove_nones(weights, initializers)
>     for w, init in zip(weights, initializers):
>         w.assign(init(w.shape, dtype=w.dtype))
> ```

And what about LSTM and BI-LSTM layers?

**Comment by phrasenmaeher** [comment]
> And what about LSTM and BI-LSTM layers?

To include these layer types, find out how and which variables are stored within a layer object. Then, add another case in the `if`-block, checking for this layer type. I imagine that you can closely follow the BatchNormalization-example in my code.
